### PR TITLE
Optimize thread ID generation

### PR DESCRIPTION
**Pull request by joboet**
By using atomics where available, thread IDs can be generated without locking while still enforcing uniqueness.

### CODE REVIEW

**Comment by 5225225** [comment]
Do we have "two threads must not have the same thread id" as a safety invariant? This allows thread id collisions.

One by simply spawning another thread, having it wrap (and set the wrapped value), then spawning another one (which will be wrapped) before the original thread manages to panic.

And another way, by simply `catch_unwind`ing the thread spawn. As far as I can see, there's nothing preventing that. The old code would just not increment the value in that case, so any attempt to spawn a thread would panic. This will write back the new value, panic, but the panic doesn't bring down the whole program.

This can be avoided by using a `fetch_update` which returns an error on overflow, so an overflowed value is never written back. Though at that point, a mutex might be faster under high contention. (Though, thread spawning doesn't feel like the kind of thing that would have high contention, you've got other costs there).

Not sure. Also, this requires 64 bit atomics to be available, which I don't think is true on all supported platforms?

**Comment by joboet** [comment]
> edit: I didn't read the code carefully enough! sorry!

No worries! GitHub somehow has problems displaying the diff; I found it hard to read even though I wrote it 😉!

File: library/std/src/thread/mod.rs [hunk 1]
```diff
@@ -1033,24 +1032,46 @@ pub struct ThreadId(NonZeroU64);
...
+
+        cfg_if::cfg_if! {
+            if #[cfg(target_has_atomic = "64")] {
+                use crate::sync::atomic::{AtomicU64, Ordering::Relaxed};
+
+                static COUNTER: AtomicU64 = AtomicU64::new(0);
+
+                let mut last = COUNTER.load(Relaxed);
+                loop {
+                    let Some(id) = last.checked_add(1) else { exhausted() };
```
Comment (joshtriplett) [inline library/std/src/thread/mod.rs:1048]:
```suggestion
                    let Some(id) = last.checked_add(1) else {
                        exhausted();
                    };
```

**Comment by joshtriplett** [comment]
LGTM; r=me with the suggestion applied.

**Comment by joboet** [comment]
@joshtriplett Done! Should I squash commits?

**Comment by joshtriplett** [comment]
@joboet Yes, please do.

**Pull request merged**